### ISSUE TITLE
Feature: S3 Object Storage

### ISSUE THREAD

**Issue by gilanghamidy** [saltcorn/saltcorn#998]
Hey all. Few days came across Saltcorn for simple no-code platform, and I find it really promising yet simple for minimal small-business use. I am planning to use Saltcorn for my use case. However, I was wondering if AWS S3 (compatible) file storage can be implemented as well.

Since no AWS S3 plugin available yet, I tried to look at the code directly, wondering if I can make a plugin for S3. But I found that the file upload mechanism is tightly coupled to the middleware and currently no access from the plugin realm.

So I decided to experiment and develop a quick-and-dirty implementation for S3 object storage. I use AWS-SDK combined with Multer to handle the multipart-form request and forward it to S3 endpoint. I also modified the route for download, serve, and delete to use S3 handler if it is configured. Additionally, I created a new settings page to enter S3 config and to enable/disable S3 file upload.

I overlayed the express-fileupload middleware to use Multer instead, because there is a middleware to forward multipart-form request to S3 directly. If we stick to use fileupload, the file will be stored first locally before being forwarded. It is too inefficient compared to passing through the multipart request to S3.

I also created some "hack" function to let the existing file upload flow intact, particularly the part where Saltcorn moves the file from temp folder to the designated local folder. Also, I modified the delete function in the File model to detach the unlink function and able to use S3 deleteObject function.

In addition, I added additional column in files table to indicate if a file is stored in S3 or locally. Therefore, enabling/disabling S3 store on the fly does not break existing files.

You can check the code here: [https://github.com/gilanghamidy/saltcorn/commit/1baf6730702a3609cb7951cd903c07c2b0dbb25f](https://github.com/gilanghamidy/saltcorn/commit/1baf6730702a3609cb7951cd903c07c2b0dbb25f) 

This is a totally quick and dirty implementation, but it is working correctly. I haven't made any test fixture for this, since I don't understand yet. I am also not sure if this work in multitenant scenario as I haven't try it yet.

Basically, I made this in just half-day, including learning to code NodeJS (I am not a JS dev, more to C++ and C#), configuring my dev machine with WSL, and reading all docs and StackOverflow to craft this. So pardon me if the implementation is not that good :)

**Comment by glutamate** [comment]
@gilanghamidy thank you for doing all this work! I will try and get this merged in the next few days with an option for storage on S3.

**Comment by gilanghamidy** [comment]
Sure! I have cleaned the code, and removed an unnecessary commit (I think) to npm-shrinkwrap.json. I'll make the official pull request so you can take a look :)

**Comment by j2l** [comment]
Wow! I don't use S3 but thank you @gilanghamidy!
Bonjour à la Belgique :)

**Comment by glutamate** [comment]
this is merged in #1001
